### PR TITLE
Add online codes

### PR DESCRIPTION
**Pull request by piowit**

### CODE REVIEW

File: app/controllers/admin/online_codes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    def export
+      @online_codes = OnlineCode.all.order(:code)
+
+      response.headers['Content-Type'] = 'text/csv'
+      response.headers['Content-Disposition'] =
+        "attachment; filename=online_codes_#{Time.zone.now.strftime('%y%m%d_%H-%M-%S')}.csv"
+      render template: 'admin/online_codes/export', handlers: [:erb], formats: [:csv]
+    end
+
+    def import
```
Comment (ciejj) [inline app/controllers/admin/online_codes_controller.rb:54]: consider wrapping it in ServiceObject

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -8,35 +8,34 @@ def index
...
-  def order_form_params
-    ofp = params.require(:order_form).permit(:reward, :delivery_method, :street, :postcode, :city, :address_id)
-    ofp[:employee] = current_employee
-    ofp
-  end
-
-  def reward
-    Reward.find(params[:reward])
+  def order_params
+    cos = params.require(:order).permit(:reward_id, address: %i[street postcode city])
```
Comment (ciejj) [inline app/controllers/orders_controller.rb]: variable name could be renamed to something more explanatory :)

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -8,35 +8,34 @@ def index
...
   end
 
   def create
-    reward = Reward.find(order_form_params[:reward])
-    if current_employee.points < reward.price
-      redirect_to rewards_path, notice: 'You have insufficient funds'
+    @reward = Reward.find(order_params[:reward_id])
+    @address = Address.new(order_params[:address])
+    @order = Order.new(reward: @reward, reward_snapshot: @reward, address_snapshot: @address, employee: @employee)
+    return redirect_to rewards_path, notice: 'You have insufficient funds' if current_employee.points < @reward.price
```
Comment (ciejj) [inline app/controllers/orders_controller.rb]: should be in service

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -8,35 +8,34 @@ def index
...
-    ofp = params.require(:order_form).permit(:reward, :delivery_method, :street, :postcode, :city, :address_id)
-    ofp[:employee] = current_employee
-    ofp
-  end
-
-  def reward
-    Reward.find(params[:reward])
+  def order_params
+    cos = params.require(:order).permit(:reward_id, address: %i[street postcode city])
+    cos[:employee] = current_employee
```
Comment (ciejj) [inline app/controllers/orders_controller.rb]: check why this is needed?

File: app/controllers/orders_controller.rb [hunk 5]
```diff
@@ -8,35 +8,34 @@ def index
...
     else
-      order_form = OrderForm.new(order_form_params)
-      if order_form.save
-        redirect_to orders_path, notice: 'Reward bought'
-      else
-        render 'new',
-               locals: { order_form: order_form, reward: reward, delivery_method: reward.delivery_method,
-                         employee: current_employee }
-      end
+      render 'new', locals: { order: @order, reward: @order.reward, employee: current_employee }
```
Comment (ciejj) [inline app/controllers/orders_controller.rb]: consider
```suggestion
      render 'new', locals: { order: create_order_service.order, reward: @order.reward, employee: current_employee }
```

File: app/mailers/delivery_order_mailer.rb [hunk 6]
```diff
@@ -4,6 +4,7 @@ class DeliveryOrderMailer < ApplicationMailer
   # send a signup email to the user, pass in the user object that   contains the user's email address
   def send_delivery_confirmation_email
     @order = params[:order]
+    @online_code = @order&.online_code&.code if @order&.reward_snapshot&.delivery_method == 'online'
```
Comment (ciejj) [inline app/mailers/delivery_order_mailer.rb]: consider moving it out of the view

File: app/models/online_code.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class OnlineCode < ApplicationRecord
+  validates :code, :reward, presence: true
+  validates :code, uniqueness: true
+
+  belongs_to :reward
+  counter_culture :reward, column_name: proc { |model| model.sent? ? nil : 'online_codes_count' }
```
Comment (ciejj) [inline app/models/online_code.rb]:
```suggestion
  counter_culture :reward, column_name: proc { |online_code| online_code.sent? ? nil : 'online_codes_count' }
```

File: app/models/online_code.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+  belongs_to :order, optional: true
+
+  scope :available, -> { where(order_id: nil) }
+
+  def sent?
+    order.present?
+  end
+
+  def self.import(file)
```
Comment (ciejj) [inline app/models/online_code.rb]: as mentioned above consider moving it to service object

File: app/models/order.rb [hunk 9]
```diff
@@ -4,6 +4,11 @@ class Order < ApplicationRecord
   serialize :reward_snapshot
   serialize :address_snapshot
 
+  has_one :online_code, dependent: :destroy
   belongs_to :employee
   belongs_to :reward
+
+  def employee_name
```
Comment (ciejj) [inline app/models/order.rb]: where do we use that?

File: app/services/create_order_service.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  end
+
+  def assign_online_code
+    @online_code = OnlineCode.available.where(reward: @reward).first
+    @online_code.update!(order: @order)
+  end
+
+  def deliver_online_code
+    @order.update!(delivered: true)
+    DeliveryOrderMailer.with(order: @order).send_delivery_confirmation_email.deliver_now
```
Comment (ciejj) [inline app/services/create_order_service.rb]: should be separate method

File: app/services/create_order_service.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    return true if @reward.delivery_method == 'online' && @reward.online_codes_count.positive?
+
+    raise StandardError, 'Not enough items in stock'
+  end
+
+  def check_funds
+    raise StandardError, 'You have insufficient funds' if @employee.points < @reward.price
+  end
+
+  def save_address
```
Comment (ciejj) [inline app/services/create_order_service.rb]: update_address?

File: app/services/create_order_service.rb [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  def initialize(params)
+    @reward = Reward.find(params[:reward_id])
+    @employee = params[:employee]
+
+    @address = @employee.address || Address.new(employee: @employee)
+    @street = params.dig(:address, :street)
+    @city = params.dig(:address, :city)
+    @postcode = params.dig(:address, :postcode)
+    
+    @errors = ActiveModel::Errors.new(self)
```
Comment (ciejj) [inline app/services/create_order_service.rb]: do we need that?

**Review by ciejj — COMMENTED** [review]
part1 done

File: app/models/order.rb (hunk 9 above)
Comment (piowit) [reply app/models/order.rb]: it was used in two views, but i changed it to `order.employee.full_name`

File: app/controllers/orders_controller.rb (hunk 4 above)
Comment (piowit) [reply app/controllers/orders_controller.rb]: moved employee from params to initialize variable

File: app/services/create_order_service.rb (hunk 12 above)
Comment (piowit) [reply app/services/create_order_service.rb]: refactored to errors array, as it's easier to handle :)

File: app/services/create_order_service.rb [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    @street = params.dig(:address, :street)
+    @city = params.dig(:address, :city)
+    @postcode = params.dig(:address, :postcode)
+
+    @errors = []
+  end
+
+  def call
+    return false unless check_items_stock
+    return false unless check_funds
```
Comment (ciejj) [inline app/services/create_order_service.rb]: sufficient_funds?

File: app/services/create_order_service.rb [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    false
+  end
+
+  def update_address
+    @address.update!(city: @city, postcode: @postcode, street: @street, last_used: Time.current)
+  end
+
+  def create_order
+    @order = Order.new(reward: @reward, reward_snapshot: @reward, address_snapshot: @address, employee: @employee)
+    @order.valid? ? @order.save! : @errors.merge!(@order)
```
Comment (ciejj) [inline app/services/create_order_service.rb]: cleanup

File: app/services/create_order_service.rb [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    true
+  rescue ActiveRecord::StatementInvalid, ActiveRecord::RecordInvalid => e
+    @errors << e.message
+    false
+  end
+
+  private
+
+  def check_items_stock
+    return true if @reward.delivery_method == 'post' && @reward.available_items.positive?
```
Comment (ciejj) [inline app/services/create_order_service.rb]: return number_of_available_items

File: app/services/create_order_service.rb [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    @address.update!(city: @city, postcode: @postcode, street: @street, last_used: Time.current)
+  end
+
+  def create_order
+    @order = Order.new(reward: @reward, reward_snapshot: @reward, address_snapshot: @address, employee: @employee)
+    @order.valid? ? @order.save! : @errors.merge!(@order)
+  end
+
+  def decrease_item_stock
+    @reward.update!(available_items: @reward.available_items - 1)
```
Comment (ciejj) [inline app/services/create_order_service.rb]: possible refactor?

File: app/services/create_order_service.rb [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    @online_code.update!(order: @order)
+  end
+
+  def deliver_email
+    if @reward.delivery_method == 'post'
+      DeliveryOrderMailer.with(order: @order).send_delivery_confirmation_email.deliver
+    else
+      DeliveryOrderMailer.with(order: @order, code: @online_code.code).send_online_code_delivery_email.deliver
+    end
+    @order.update!(delivered: true) if @order.reward_snapshot.delivery_method == 'online'
```
Comment (ciejj) [inline app/services/create_order_service.rb:74]: @order.deliver! ?

File: app/services/import_online_codes_csv_service.rb [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+    true
+  rescue ActiveRecord::StatementInvalid, ActiveRecord::RecordInvalid => e
+    @errors << e.message
+    false
+  end
+
+  private
+
+  def check_file
```
Comment (ciejj) [inline app/services/import_online_codes_csv_service.rb]: file_valid?, readable_csv?

File: app/views/admin/orders/export.csv copy.erb [hunk 19]
```diff
@@ -0,0 +1,2 @@
+<% headers = ['slug', 'title', 'description', 'price', 'delivery_method'] -%><%= CSV.generate_line headers -%>
```
Comment (ciejj) [inline app/views/admin/orders/export.csv copy.erb]: ?

File: spec/models/online_code_spec.rb [hunk 20]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe OnlineCode, type: :model do
+  let(:online_code) { create(:online_code) }
```
Comment (ciejj) [inline spec/models/online_code_spec.rb]: let's try not using let :) ?

File: spec/services/create_order_service_spec.rb [hunk 21]
```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe CreateOrderService do
+  let!(:employee) { create(:employee) }
+
+  context 'when buying postal reward' do
+    let!(:reward_post) { create(:reward, price: 1, delivery_method: 'post', available_items: 100) }
```
Comment (ciejj) [inline spec/services/create_order_service_spec.rb]: traits?

File: app/services/create_order_service.rb [hunk 22]
```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+class CreateOrderService
+  attr_reader :errors, :order, :reward, :address
+
+  def initialize(params, employee)
```
Comment (ciejj) [inline app/services/create_order_service.rb]:
```suggestion
  def initialize(params:, employee:)
```

File: spec/services/create_order_service_spec.rb [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+
+    before do
+      create(:kudo, receiver: employee) # add funds
+    end
+
+    context 'when all attributes are present' do
+      it 'is valid' do
+        params = { address: attributes_for(:address), reward_id: reward_post.id }
+        create_order_service = described_class.new(params, employee)
```
Comment (ciejj) [inline spec/services/create_order_service_spec.rb]: subject?

File: spec/services/create_order_service_spec.rb [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+
+      it 'delivers confirmation email' do
+        params = { address: attributes_for(:address), reward_id: reward_post.id }
+        create_order_service = described_class.new(params, employee)
+        expect { create_order_service.call }.to change { ActionMailer::Base.deliveries.count }.by(1)
+      end
+    end
+
+    context 'when attributes are missing' do
+      it 'is invalid without street' do
```
Comment (ciejj) [inline spec/services/create_order_service_spec.rb]: it return false when address is invalid

File: spec/services/create_order_service_spec.rb [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+
+      it 'is invalid without city' do
+        params = { address: attributes_for(:address, city: nil), reward_id: reward_post.id }
+        create_order_service = described_class.new(params, employee)
+        expect(create_order_service.call).to be false
+        expect(create_order_service.errors.to_s).to include "City can't be blank"
+      end
+    end
+
+    it 'is invalid when not enough funds' do
```
Comment (ciejj) [inline spec/services/create_order_service_spec.rb]: returns false

**Review by ciejj — COMMENTED** [review]
part 2

File: app/services/create_order_service.rb (hunk 17 above)
Comment (piowit) [reply app/services/create_order_service.rb:74]: this is just boolean in db, not enum

File: spec/services/create_order_service_spec.rb [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+    it 'delivers online_code via email after purchase' do
+      employee = create(:employee)
+      create(:kudo, receiver: employee) # add funds
+      reward_online = create(:reward, price: 1, delivery_method: 'online')
+      create(:online_code, reward: reward_online)
+      params = { employee: employee, reward_id: reward_online.id }
+      create_order_service = CreateOrderService.new(params: params, employee: employee)
+      expect { create_order_service.call }.to change { ActionMailer::Base.deliveries.count }.by(1)
+      expect(Order.last.delivered).to eq(true)
+      expect(ActionMailer::Base.deliveries.last.body.to_s).to include OnlineCode.first.code
```
Comment (ciejj) [inline spec/services/create_order_service_spec.rb]: lets check if proper email has been sent

File: spec/services/create_order_service_spec.rb [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+    it 'returns false when there is no available online_codes for ' do
+      employee = create(:employee)
+      create(:kudo, receiver: employee) # add funds
+      reward_online = create(:reward, price: 1, delivery_method: 'online')
+      params = { employee: employee, reward_id: reward_online.id }
+      create_order_service = CreateOrderService.new(params: params, employee: employee)
+      expect(create_order_service.call).to be false # there were no online_codes present
+    end
+
+    it 'returns false when emoployee dosnt have funds for reward' do
```
Comment (ciejj) [inline spec/services/create_order_service_spec.rb]: typo

**Review by ciejj — APPROVED** [review]
👍

**Pull request merged**